client/iotcore/client.go

```diff
@@ -0,0 +1,315 @@
+// Copyright 2022 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package iotcore
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"flag"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/iot"
+)
+
+var (
+	ErrDeviceNotFound = errors.New("device not found")
+)
+
+//nolint:lll
+//go:generate ../../utils/mockgen.sh
+type Client interface {
+	GetDevice(ctx context.Context, sess *session.Session, deviceID string) (*Device, error)
+	UpsertDevice(ctx context.Context, sess *session.Session, deviceID string, device *Device, policy string) (*Device, error)
+	DeleteDevice(ctx context.Context, sess *session.Session, deviceID string) error
+}
+
+type client struct{}
+
+func NewClient() Client {
+	return &client{}
+}
+
+func (c *client) GetDevice(
+	ctx context.Context,
+	sess *session.Session,
+	deviceID string,
+) (*Device, error) {
+	svc := iot.New(sess)
+
+	resp, err := svc.DescribeThingWithContext(ctx,
+		&iot.DescribeThingInput{
+			ThingName: aws.String(deviceID),
+		})
+
+	var device *Device
+
+	if err == nil {
```

Comment:
Why? This is part of the SDK contract: you either get a response, or an error.

Reply:
since I approved we can safely argue now :)
call me an old timer, but no matter what, I believe in always checking the pointer before using it.
Re: contract: this is not java, which had contracts as it was founded on standards which you implemented. in here it is just a library that someone wrote and there is no way to enforce nor to expect anything, since it does not comply to standard you can refer to. and I know we do it all over the code base. I propose to work on a document called: _Mender coding standard_. much of the reviews ping-pong can be avoided if we had guidelines.